src/Nethermind/Nethermind.Evm/OverridableEnv/IOverridableEnv.cs

```diff
@@ -0,0 +1,37 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Generic;
+using Autofac.Core;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+
+namespace Nethermind.Evm.OverridableEnv;
+
+/// <summary>
+/// An <see cref="IOverridableEnv"/> is an environment where the world state and or the code repository can be overridden.
+/// It is an <see cref="IModule"/> that can be used to configure an autofac child lifetime to provide the necessary components
+/// that reflect the override. Any components within that lifetime should run between the build and dispose of the returned
+/// disposable or there may be memory leak.
+/// </summary>
+public interface IOverridableEnv : IModule
+{
+    IDisposable Build(Hash256 stateRoot);
+    IDisposable BuildAndOverride(BlockHeader header, Dictionary<Address, AccountOverride>? stateOverride);
+}
+
+/// <summary>
+/// I small wrapper around <see cref="IOverridableEnv"/> to help prevent accidentally using components that rely on
+/// the overridden env outside of the env scope. Ideally, no other components from the child lifetime is extracted
+/// aside from <see cref="IOverridableEnv{T}"/>. To use any components such as <see cref="ITransactionProcessor"/>, set the
+/// <see cref="T"/> here to that components or a wrapper if many component, then call one of the method here to get it.
+/// Always dispose the scope when finished.
+/// </summary>
+/// <typeparam name="T"></typeparam>
+public interface IOverridableEnv<T>
+{
+    IDisposable Build(Hash256 stateRoot, out T component);
+    IDisposable BuildAndOverride(BlockHeader header, out T component);
+    IDisposable BuildAndOverride(BlockHeader header, Dictionary<Address, AccountOverride>? stateOverride, out T component);
+}
```

Comment:
```suggestion
public interface IOverridableEnv<T> where T : IDisposable
{
    IDisposable Build(Hash256 stateRoot, out T component);
    IDisposable BuildAndOverride(BlockHeader header, out T component);
    T BuildAndOverride(BlockHeader header, Dictionary<Address, AccountOverride>? stateOverride);}
```

Reply:
Can T be refstruct? That way it must stay in  call stack. But then it make it hard to use async.

Reply:
I think you can allow for ref struct but you can't constraint to it `where T : allows ref struct`
https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/constraints-on-type-parameters